### ISSUE TITLE
Suppress irrelevant console output from automatic capabilities test on node startup

### ISSUE THREAD

**Issue by cyberj0g** [livepeer/go-livepeer#2364]
## Abstract
With a recent update, when transcoder node starts, it tries to transcode a few test segments multiple times, to experimentally determine, which subset of capabilities this node supports. It produces a lot of output to the console, including errors, which may be confusing for users. We need to suppress this output.
<details>
  <summary>Sample output</summary>

```
I0413 03:13:01.089673   25353 livepeer.go:275] ***Livepeer is running on the offchain network***
I0413 03:13:01.090534   25353 db.go:363] Initialized DB node
I0413 03:13:01.090567   25353 livepeer.go:332] Netint devices: [0]
I0413 03:13:01.090583   25353 livepeer.go:336] Transcoding on these Netint devices: [0]
fname TEST2/testseg.tempfile
[h264_ni_enc @ 0x5656580] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x5656580] dts offset: 2, gop_offset_count: 0
[h264_ni_enc @ 0x56477c0] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x56477c0] dts offset: 2, gop_offset_count: 0
[h264_ni_enc @ 0x57887c0] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x57887c0] dts offset: 2, gop_offset_count: 0
[h264_ni_enc @ 0x581ef00] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x581ef00] dts offset: 2, gop_offset_count: 0
ni_device_session_flush p_ctx->internal_flush: 1
xcoder_config_instance_flush(): enter
xcoder_config_instance_flush(): send
xcoder_config_instance_flush(): exit
[h264_ni_dec @ 0x5680080] ff_decode_get_packet NO need to send sentinel for now !, s->api_ctx.internal_flush=1
Decoder HW[0] INST[127]-average usage:0%
Encoder HW[0] INST[127]-average usage:1%
Encoder HW[1] INST[127]-average usage:1%
Encoder HW[2] INST[127]-average usage:1%
Encoder HW[3] INST[127]-average usage:1%
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5642900] Codec type or id mismatches
ERROR: decoder.c:275] Unable to open video decoder : Invalid argument
ERROR: decoder.c:302] Unable to open video decoder : Invalid argument
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.301980   25353 ffmpeg.go:790] Transcoder Return : Invalid argument
I0413 03:13:01.302022   25353 transcoder.go:266] optional capability "HEVC decode" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5653fc0] Codec type or id mismatches
ERROR: decoder.c:275] Unable to open video decoder : Invalid argument
ERROR: decoder.c:302] Unable to open video decoder : Invalid argument
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.303099   25353 ffmpeg.go:790] Transcoder Return : Invalid argument
fname TEST2/testseg.tempfile
[h265_ni_enc @ 0x59b6540] Session state: 0 allocate frame fifo.
[h265_ni_enc @ 0x59b6540] dts offset: 2, gop_offset_count: 0
[h265_ni_enc @ 0x57b1c40] Session state: 0 allocate frame fifo.
[h265_ni_enc @ 0x57b1c40] dts offset: 2, gop_offset_count: 0
[h265_ni_enc @ 0x564ce80] Session state: 0 allocate frame fifo.
[h265_ni_enc @ 0x564ce80] dts offset: 2, gop_offset_count: 0
[h265_ni_enc @ 0x59b1280] Session state: 0 allocate frame fifo.
[h265_ni_enc @ 0x59b1280] dts offset: 2, gop_offset_count: 0
ni_device_session_flush p_ctx->internal_flush: 1
xcoder_config_instance_flush(): enter
xcoder_config_instance_flush(): send
xcoder_config_instance_flush(): exit
[h264_ni_dec @ 0x5809880] ff_decode_get_packet NO need to send sentinel for now !, s->api_ctx.internal_flush=1
Decoder HW[0] INST[127]-average usage:0%
Encoder HW[0] INST[127]-average usage:1%
Encoder HW[1] INST[127]-average usage:1%
Encoder HW[2] INST[127]-average usage:1%
Encoder HW[3] INST[127]-average usage:1%
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5809880] Codec type or id mismatches
ERROR: decoder.c:275] Unable to open video decoder : Invalid argument
ERROR: decoder.c:302] Unable to open video decoder : Invalid argument
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.511899   25353 ffmpeg.go:790] Transcoder Return : Invalid argument
I0413 03:13:01.511924   25353 transcoder.go:266] optional capability "VP8 decode" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5809880] Codec type or id mismatches
ERROR: decoder.c:275] Unable to open video decoder : Invalid argument
ERROR: decoder.c:302] Unable to open video decoder : Invalid argument
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.512303   25353 ffmpeg.go:790] Transcoder Return : Invalid argument
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5809880] Codec type or id mismatches
ERROR: decoder.c:275] Unable to open video decoder : Invalid argument
ERROR: decoder.c:302] Unable to open video decoder : Invalid argument
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.512987   25353 ffmpeg.go:790] Transcoder Return : Invalid argument
I0413 03:13:01.513004   25353 transcoder.go:266] optional capability "VP9 decode" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5809880] Codec type or id mismatches
ERROR: decoder.c:275] Unable to open video decoder : Invalid argument
ERROR: decoder.c:302] Unable to open video decoder : Invalid argument
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.513461   25353 ffmpeg.go:790] Transcoder Return : Invalid argument
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5809880] Error: pixel format yuv444p not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.514892   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
I0413 03:13:01.514911   25353 transcoder.go:266] optional capability "H264 Decode YUV444 8-bit" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5685400] Error: pixel format yuv444p not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.515878   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5685400] Error: pixel format yuv422p not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.516970   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
I0413 03:13:01.516987   25353 transcoder.go:266] optional capability "H264 Decode YUV422 8-bit" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x55eac80] Error: pixel format yuv422p not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.517910   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5685400] Error: pixel format yuv444p10le not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.519113   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
I0413 03:13:01.519131   25353 transcoder.go:266] optional capability "H264 Decode YUV444 10-bit" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x5685400] Error: pixel format yuv444p10le not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.520115   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x55eac80] Error: pixel format yuv422p10le not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.521263   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
I0413 03:13:01.521281   25353 transcoder.go:266] optional capability "H264 Decode YUV422 10-bit" is not supported on device 0, see other error messages for details
fname TEST2/testseg.tempfile
[h264_ni_dec @ 0x566e180] Error: pixel format yuv422p10le not supported.
ERROR: decoder.c:275] Unable to open video decoder : Invalid data found when processing input
ERROR: decoder.c:302] Unable to open video decoder : Invalid data found when processing input
decoder.c:314] Freeing input based on OPEN INPUT error
E0413 03:13:01.522318   25353 ffmpeg.go:790] Transcoder Return : Invalid data found when processing input
fname TEST2/testseg.tempfile
[h264_ni_enc @ 0x56670c0] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x56670c0] dts offset: 2, gop_offset_count: 0
[h264_ni_enc @ 0x58575c0] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x58575c0] dts offset: 2, gop_offset_count: 0
[h264_ni_enc @ 0x5663d00] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x5663d00] dts offset: 2, gop_offset_count: 0
[h264_ni_enc @ 0x56721c0] Session state: 0 allocate frame fifo.
[h264_ni_enc @ 0x56721c0] dts offset: 2, gop_offset_count: 0
ni_device_session_flush p_ctx->internal_flush: 1
xcoder_config_instance_flush(): enter
xcoder_config_instance_flush(): send
xcoder_config_instance_flush(): exit
[h264_ni_dec @ 0x55eac80] ff_decode_get_packet NO need to send sentinel for now !, s->api_ctx.internal_flush=1
Decoder HW[0] INST[127]-average usage:0%
Encoder HW[0] INST[127]-average usage:0%
Encoder HW[1] INST[127]-average usage:1%
Encoder HW[2] INST[127]-average usage:1%
Encoder HW[3] INST[127]-average usage:1%
I0413 03:13:01.731477   25353 livepeer.go:415] ***Livepeer is in off-chain mode***
I0413 03:13:01.731658   25353 livepeer.go:1056] ***Livepeer Running in Orchestrator Mode***
I0413 03:13:01.731782   25353 webserver.go:72] CLI server listening on 127.0.0.1:7936
I0413 03:13:01.731733   25353 mediaserver.go:188] Transcode Job Type: [{P240p30fps4x3 600k 30 0 320x240 4:3 0 0 0s 0 0 0} {P360p30fps16x9 1200k 30 0 640x360 16:9 0 0 0s 0 0 0}]
I0413 03:13:01.731803   25353 cert.go:83] Private key and cert not found. Generating
I0413 03:13:01.732091   25353 cert.go:22] Generating cert for 127.0.0.1
I0413 03:13:01.732462   25353 rpc.go:197] Listening for RPC on :8935
I0413 03:13:01.732575   25353 livepeer.go:1083] Orchestrator server shut down
I0413 03:13:01.732589   25353 db.go:368] Closing DB
```

</details>

## Motivation

Above.

## Proposed Solution

We could easily suppress `stdout` and `stderr` output in Golang, similar to [this](https://stackoverflow.com/a/58720235/4939144). However, most of above output is produced by C code of `lpms` and Ffmpeg, and it will require a very hacky and non-cross platform approach to suppress this, something like hooking `write()` syscall.

I think, a good solution would be to run capability test as a separate subprocess:

1. Add `-testCapabilities` argument to `livepeer`
2. Exit immediately after capabilities test, if `-testCapabilities` is specified
3. Parent process will create child process with same arguments + `-testCapabilities`
4. Parent process will buffer `stdout` and `stderr` of child process and output it only on high verbosity levels, or after getting non-zero exit code. It may display capability-related lines depending on verbosity.
5. Parent process will exit with corresponding error message on receiving non-zero exit code from child process.
6. Parent process should implement time out (of less than 60 secs) on child process running capabilities test.

Let me know what you think @AlexKordic @leszko @thomshutt @yondonfu.

**Comment by leszko** [comment]
Thanks for the feature spec @cyberj0g

Some thoughts:

1. Maybe it's possible to create a C-based wrapper and redirect stdout there if it's not possible in Golang? I found a similar issue with Python, you can find some solution ideas [here](https://stackoverflow.com/questions/5081657/how-do-i-prevent-a-c-shared-library-to-print-on-stdout-in-python).

2. To me, spawning a separate process only to suppress the logs sounds a little like an overhead, at least that's what I think at the first glance. There is some simplicity in having just one Livepeer process, for example, you start Livepeer and all of the sudden you terminate the process, I guess your capabilities test will keep on running.

In any case, no blocker from my side, but I'd consider if we really don't have any other option to suppress the stdout logs created by libraries.

**Comment by thomshutt** [comment]
I think that one nice thing that having a separate process might give us is to have some sort of fallback behavior in Golang in the case where the capabilities test fails, without completely failing to start up. Outside of that though I'd be inclined to agree with @leszko that we should in general be trying to capture the logs in the calling process if possible.

**Comment by AlexKordic** [comment]
@cyberj0g Can we use [av_log_set_callback()](https://github.com/livepeer/FFmpeg/blob/livepeer/libavutil/log.h#L321) to filter unwanted output?

**Comment by cyberj0g** [comment]
Other subprocess approach benefits are:
* minimal modifications to the code: no need to modify logging, where it can't be turned off programmatically now
* guarantee of resource cleanup, because process exits (say, we have some library for custom hardware, which doesn't release memory properly in case of specific capability test failure)

But I see the simplicity argument. We actually can suppress 2/3 of excessive log messages with [av_log_set_level](https://ffmpeg.org/doxygen/2.3/group__lavu__log.html#ga1fd32c74db581e3e2e7f35d277bb1e24), and disabling glog output, this would leave only `printf` outputs from Netint code. I'll look into how well closing stdout/stderr works with Ffmpeg, thanks for the tip @leszko.

**Comment by AlexKordic** [comment]
Will subprocess change prevent simple debugging of C code?

**Comment by cyberj0g** [comment]
@AlexKordic my intuition is that it should not hurt debugging experience, most debuggers attach to child processes automatically. But it needs to be proven for Go -> C case.

**Comment by yondonfu** [comment]
Closed by #2396
